### ISSUE TITLE
Reach Out To Other Open Standards

### ISSUE THREAD

**Issue by shaggygi** [Azure/IoTPlugandPlay#10]
This new data modeling language is interesting and I can see potential.  However, I have to wonder would it be better to reach out to other open standards groups instead of reinventing the wheel.  I'm sure there are many in the wild, but I sort of like the following example.

There is a BACnet ([Building Automation and Control Networks](http://www.bacnet.org/)) abstract model now released/published as Annex Y.  Here is a [link to early draft](http://www.bacnet.org/Addenda/Add-135-2012am-PPR1-draft-26_chair_approved.pdf).  Understanding this is within the umbrella of BACnet (which has to do much about IoT), but you can actually use the model to describe almost anything including devices, objects, people, buildings, protocols, etc..  In addition, the specification generally explains the model and includes examples of how to format in XML and JSON.

Just curious.  Keep up the good work and I look forward to future releases.

**Comment by briancr-ms** [comment]
Thanks for your comments about other standards. You gave BACnet as an example. Which modeling standards do you currently use and which ones would be most important to you to see supported?

**Comment by shaggygi** [comment]
Thanks for the reply.

I have not really used a modeling standard, at least in any meaningful way.  If I did, I would seriously consider advocating the BACnet abstract model.  And as mentioned, even though that is under the BACnet spec, it is outlined in a way that could model various topics and not just a protocol.  You can see a Java concept implementation [here](https://github.com/ampii/ampii-server).

There is also another tagging standards (and is also related to building automation) called [Project Haystack](https://marketing.project-haystack.org/blog/haystack-4-is-coming-what-it-is-and-why-it-matters).

**Comment by shaggygi** [comment]
I'm going to try and dust off my notes and create an example comparing it to one of the devices in the current catalog.  Give me a few days.

**Comment by gizmochief7** [comment]
While I do agree that other open standards should be brought into the mix. I don't necessarily think Microsoft is reinventing  the wheel any more than those open standards are reinventing the idea of UPnP. 

If I was to suggest other standards that I currently use, I would love to see full integration with `OpenAPI 3.0` aka Swagger. The primary reason for this is the vast ecosystem it has and abundance of tooling that exist that can autogenerate both server and client code in many of the most popular languages with only the OpenAPI.json specification file as input. This reenforces the "API first" mentality. Another powerful feature of OpenAPI ecosystem is the Swagger-UI which can generate an API specification webpage to visually and intuitively introduce your API to other developers.

Other non API standards but large IoT open source projects I'd love to see this integration with would be Home Assistant platform and Tasmota

Joking side note: Microsoft should just buy Espressif for the ESP32, just saying

**Comment by shaggygi** [comment]
Side note... would it be possible to use this format with gRPC?

**Comment by ChrisGMsft** [comment]
We would appreciate more input from others in the community regarding alignment with existing standards. We'll have a more formal process for this, but in the meantime we'd encourage others to add their vote. Thanks!

**Comment by imrankhan1984** [comment]
I agree with the open standards. Also why not look into standard ontology works like Smart Appliance REFerence (SAREF) ontology from Etsi.org and Semantic Sensor Network (SSN) from W3C and reuse well-defined concepts.

**Comment by imrankhan1984** [comment]
A good place to start is [Web of Things (WoT) Thing Description](https://www.w3.org/TR/wot-thing-description/). They support inclusion of external vocabularies like SAREF, IoT-Schema.org and SSN into their description model.

**Comment by briancr-ms** [comment]
Thanks for all the comments and ideas on this thread! We have been looking at these (and other) vocabularies and want to enable them to be used with IoT plug & play. Please keep the feedback on this thread coming--we want to hear what vocabularies you're interested in and any other standards you're interested in.

**Comment by shaggygi** [comment]
@briancr-ms Thanks for the update.  If possible, it would be nice to see some type of roadmap or current thinking... even if it isn't detailed and not completely understood where this project is headed.

As mentioned before, I'm interested in this space.  However, coming from my perspective being an outsider, it is a little worrisome to invest based on the little repo traffic since the announcement.

Are there any updates to DTDL, early APIs to tinker with, possible dev boards/files and samples to evaluate, etc.?

Being involved with the .NET IoT project, I can see where we could benefit from this, but not sure if the timing is right yet.

Thanks again.

**Comment by ChrisGMsft** [comment]
@shaggygi we do have a preview coming, at which point you'll be able to evaluate the APIs and samples. Stay tuned!

**Comment by shaggygi** [comment]
Another area to see how it could fit is the Open Data Initiative and [Common Data Model](https://github.com/microsoft/CDM/issues) that Microsoft is driving.

**Comment by psereiko** [comment]
The process automation industry has used a standard called EDDL for many many years. Here's a link https://www.amazon.de/EDDL-Electronic-Device-Description-Language/dp/3835632434
EDDL is also a IEC standard IEC 61804-3.

**Comment by shaunco** [comment]
Just to add to the mix, with Microsoft being a platinum member of the Open Connectivity Foundation, it would also make sense to use oneIoTa resource models instead of creating an entirely new set of device data models.
https://oneiota.org

**Comment by eoursel** [comment]
I would raise the same question about OPC UA (IEC62541)  Information Models especially devices models based on IEC62541-100. Is someone is working to map the OPC UA Metamodel to DTDL? Any progress on the subject? 

Regards,

**Comment by mrentsch65** [comment]
In this context I would like to point to [AutomationML](https://de.wikipedia.org/wiki/AutomationML), they provide the most mature ecosystem: 
https://www.automationml.org/o.red.c/dateien.html

**Comment by Manu3756** [comment]
Also the [Asset Administration Shell](https://www.plattform-i40.de/PI40/Redaktion/DE/Downloads/Publikation/Details-of-the-Asset-Administration-Shell-Part1.html) by [Plattform Industrie 4.0](https://www.plattform-i40.de/PI40/Navigation/DE/Home/home.html) should be mentioned here.

**Comment by kolotu** [comment]
I would also suggest to add the semantic IoT modelling project [Eclipse Vorto](https://github.com/eclipse/vorto) here. The concept is very similar - the [official repository](https://vorto.eclipse.org) for Vorto models provides some tooling around the models.

**Comment by OR13** [comment]
I suggest getting someone to engage:

- https://github.com/w3c/did-core
- https://github.com/w3c/vc-data-model/

**Comment by juusoautiosalo** [comment]
A recent article ["Digital Twin and Internet of Things—Current Standards Landscape](https://doi.org/10.3390/app10186519)" might provide useful information for this issue.

Table 1 in Section "4. Summary" gives a quick comparison overview on these standards:
* [Asset Administration Shell (AAS)](https://www.plattform-i40.de/PI40/Redaktion/EN/Downloads/Publikation/Details_of_the_Asset_Administration_Shell_Part1_V3.pdf?__blob=publicationFile&v=5) (Dev starter kit [here](https://www.plattform-i40.de/PI40/Redaktion/EN/Downloads/Publikation/VWSiD%20V2.0.pdf?__blob=publicationFile&v=2))
* Digital Twin Definition Language (DTDL)
* [Next Generation Service Interfaces-Linked Data (NGSI-LD) API](https://www.etsi.org/deliver/etsi_gs/CIM/001_099/009/01.02.02_60/gs_CIM009v010202p.pdf)
* [Open Data Protocol (OData)](https://www.odata.org/)
* [SensorThings API (STA)](https://www.ogc.org/standards/sensorthings)
* [Web of Things Thing Description (WoT TD)](https://www.w3.org/TR/wot-thing-description/)

**Comment by briancr-ms** [comment]
Thanks for the pointer to this article. It's a good look at the landscape and includes many of the standards we have been looking at and working with.

**Comment by ahernandez-allegion** [comment]
See other related issues: 

https://github.com/Azure/opendigitaltwins-dtdl/issues/20
https://github.com/Azure/opendigitaltwins-dtdl/issues/43

**Comment by psereiko** [comment]
FieldComm Group has recently released the Process Automation - Device Information Model (PA-DIM) specification based on OPC UA information modeling. You can review the spec here: 
[https://library.fieldcommgroup.org/10098/TS10098/1.0/#page=1](url)

**Comment by oising** [comment]
Hah. The comments on this thread are all you need to know about how easy it is for people to agree on a standard.

**Comment by shaggygi** [comment]
Yup, sad i even created the issue.  Muting now.
